scuttlebutt-mux/src/main/java/net/consensys/cava/scuttlebutt/mux/RPCHandler.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package net.consensys.cava.scuttlebutt.mux;
+
+import net.consensys.cava.bytes.Bytes;
+import net.consensys.cava.scuttlebutt.handshake.vertx.ClientHandler;
+import net.consensys.cava.scuttlebutt.rpc.RPCCodec;
+import net.consensys.cava.scuttlebutt.rpc.RPCFlag;
+import net.consensys.cava.scuttlebutt.rpc.RPCMessage;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+
+public class RPCHandler implements Multiplexer, ClientHandler {
+
+  private final Consumer<Bytes> messageSender;
+  private Map<Integer, CompletableFuture<RPCMessage>> awaitingAsyncResponse = new HashMap<>();
+
+  private Map<Integer, ScuttlebuttStreamHandler> streams = new HashMap<>();
+
+  /**
+   *
+   * @param messageSender sends the request to the server
+   */
+  public RPCHandler(Consumer<Bytes> messageSender, Runnable terminationFn) {
+    this.messageSender = messageSender;
+  }
+
+
+  @Override
+  public CompletableFuture<RPCMessage> makeAsyncRequest(RPCMessage request) {
+    CompletableFuture<RPCMessage> future = new CompletableFuture<>();
```

Comment:
is there a need for a future here? It is not marked complete.

Reply:
These futures are resolved in the `receivedMessage` method. in this `RPCHandler` class. 

When a request is made, they're stored in a map of 'active' requests which haven't been resolved by the peer we're making the request to yet. Once a response is received and the `receivedMessage` method is called - the map of active requests is looked up by request number, and the future is resolved.

Reply:
OK got it. Can we use AsyncResult instead? We developed kind of our own as CompletableFuture API is bit hairy.

Reply:
Sure, no problem :)